common/src/main/java/io/netty/util/concurrent/DefaultEventExecutorChooserFactory.java

```diff
@@ -53,7 +69,10 @@ private static boolean isPowerOfTwo(int val) {
 
         @Override
         public EventExecutor next() {
-            return executors[idx.getAndIncrement() & executors.length - 1];
+            final AtomicInteger idx = this.idx;
+            final EventExecutor[] executors = this.executors;
+            final int executorsMask = executors.length - 1;
+            return executors[idx.getAndIncrement() & executorsMask];
```

Comment:
Exactly, a similar pattern is been used in the many concurrent primitives of `java.util.concurrent.*` classes; copied from `ForkJoinPool` code:
```
     * Memory ordering relies mainly on Unsafe intrinsics that carry
     * the further responsibility of explicitly performing null- and
     * bounds- checks otherwise carried out implicitly by JVMs.  This
     * can be awkward and ugly, but also reflects the need to control
     * outcomes across the unusual cases that arise in very racy code
     * with very few invariants. So these explicit checks would exist
     * in some form anyway.  All fields are read into locals before
     * use, and null-checked if they are references.  This is usually
     * done in a "C"-like style of listing declarations at the heads
     * of methods or blocks, and using inline assignments on first
     * encounter.  Array bounds-checks are usually performed by
     * masking with array.length-1, which relies on the invariant that
     * these arrays are created with positive lengths, which is itself
     * paranoically checked.
```
TLDR; primitives like `getAndIncrement()` can force instance fields to be read several times and bounds checks to happen more then necessary: on hot paths it doesn't cost nothing (if not ugliness of code) to use this declaration style.

Reply:
I guess I don't follow the direct equivalence between this what's described there which appears to relate to the need for explicit bounds/null checks due to use of `Unsafe`, I am probably still missing the subtlety! Is it that the non-plain `getAndIncrement` invocation could happen between reads of the `executors` field? I would have thought the most that could help is a local var for that once since it's the only one read more than once?

Reply:
I've probably misjudged the code I've seen and I can revert this change in this specific case. The original idea is to control how many times  "executors" has been fetched and null checked (in this case, by the JVM) due to the ordering imposed by getAndIncrement.
I see now that "executors" will be used right after getAndIncrement ie it shouldn't fetched twice.